### ISSUE TITLE
Bug: colors in line charts are often very similar

### ISSUE THREAD

**Issue by lucasrodes** [owid/owid-grapher#3898]
## Description
Line colors in line charts colors are sometimes very similar depending on the entities selected.
 
## Expected behavior
All lines in the chart should have colors that are easy to distinguish.

## Steps to reproduce

1. Go to https://ourworldindata.org/grapher/excess-mortality-p-scores-projected-baseline
2. Follow the steps displayed in the first video below

## Additional context and screenshots

After playing with selecting and unselecting entities in [this chart](https://ourworldindata.org/grapher/excess-mortality-p-scores-projected-baseline), I realized that the lines were not very distinct. I am a bit confused by the rules for the Grapher when choosing colors.

https://github.com/user-attachments/assets/8bbcf9da-291a-483e-aa60-41b8aa8dffcd

If one refresh, the chart is loaded with distinguishable colors (second video).

https://github.com/user-attachments/assets/4b7b7cb6-bdae-49e1-a793-f2f67192f71b

I think this often leads to DI charts with similar line colors.

## Environment

For desktop, provide:

-   OS: MacOS
-   Browser: Arc
-   Version: 1.57.1 (Chromium Engine Version 128.0.6613.85)

**Comment by toni-sharpe** [comment]
**This isn't a high priority comment**. But I hope it is useful to you. If not and these sorts of comment are wasting your time or just annoying, ask me to stop, I won't be offended. I've tried to format it so it's not overwhelming but so there's some deeper dives possible for interested folk.

I'm assuming low knowledge and sticking at that pace, it's not directed at any one person's decision or work. A search of the grapher repo returns "192 matches across 51 files" for `rgb` and "313 matches across 55 files" for `: #` indicating a CSS hex declaration. 54 of those are .scss files. `hsl` is used 24 times in 4 places.

## HSL in a layman style

**tl;dr**: HSL presents colour as a set of values designed for people, unlike HEX, designed for computers. It's syntactic sugar really, but improves the DX drastically. **H**ue is the colour of visible light, **s**aturation is how much of it there is (fade to grey) and **l**ightness (or **l**uminescence) is how much light there is (fade to black).

Also it shows how the values can be manipulated with simple to understand arithmetic. Images taken from [GIMP](https://www.gimp.org/) are used to demonstrate this, and there are some links to [CSS tricks](https://css-tricks.com/yay-for-hsla/) and [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/hsl) which get into the technical detail.

<details>
  <summary>HSL, how is works, etc. - large section with images</summary>
  
=====================================

### HSL Definition:

**Hue**: is a wheel of colour that rolls round from red, through orange, to yellow, then green, blue, purple and finally back to red. Typically it runs through 360 degrees, like a compass. Note that [there is no pink](https://www.youtube.com/watch?v=S9dqJRyk0YM). Pink is really light red, you'll notice on some artists paints that the pinks are in the red range.

**Saturation**: is the amount of hue in the result. Usually a %. Set to `100` it's as much colour as possible, set to `0` it's grey-scale.

**Lightness / Luminescence**: Two words that mean roughly the same and start with L. Again, usually a % with `0` being absolute darkness, the absence of light. Technically this hasn't been achieved by humans although there are paints[1] that absorb 99.9% of light. `100` is the opposite. The result is a scale from black to white, which goes through all levels of the selected hue, at the selected saturation.

HSL is usually much simpler for people to understand, whereas mixtures of RGB are more computer focused. Monitors use a mixture of R, G and B to produce the range of colours, which we know from hex codes in CSS, whereas HSL puts the colour into one place which we understand because it is closer to what we see then uses saturation and lightness, again, a closer match to our experience with eyes, assuming sightedness.

It's also much easier to do arithmetic with.

[1] now isn't the time to get into why paint uses yellow and not green as a primary colour.

### In CSS

I'll let the excellent CSS Tricks site explain [how it works in CSS style sheets](https://css-tricks.com/yay-for-hsla/).

And [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/hsl), just because they're great.

CSS also has an alpha property is HSLA, that defines opaqueness.

### An example

In these examples look at the bottom left for "Current" colour.

![Screenshot 2024-08-28 at 02 07 11](https://github.com/user-attachments/assets/e101e2e8-06d5-4aff-bde7-c79175a57985)

### A hue change

You can see how the S and L portions are the same but the Hue selection is moved.

<img width="1127" alt="Screenshot 2024-07-28 at 16 01 18" src="https://github.com/user-attachments/assets/cf33e9d0-7263-472c-bf7d-f51d1a4d5efc">

### Grey-scale

Here we see the exact same Hue and Lightness as the previous example, but the removal of all saturation results in grey. This grey would be the same regardless of hue.

<img width="1128" alt="Screenshot 2024-07-28 at 15 48 06" src="https://github.com/user-attachments/assets/b7b2d13c-8ec7-4c53-bd5f-111bdc432091">

### Black

Here's the same hue again, this time with no lightness, resulting in black. The same applies where this occurs with any hue.

<img width="1119" alt="Screenshot 2024-07-28 at 15 48 34" src="https://github.com/user-attachments/assets/404df078-bfeb-4981-b2b3-2392c79e9312">

### Final note

HSL is really syntactic sugar for referring to light from a luminescent source, in a computer it's converted into RGB for display; this is different from light reflected off an object (like something painted) which is why RGB are primary on screens, and Red, Blue and Yellow are primary in paints.

I'll leave colour there for now, hope this explanation helps. The topic can take a life's work and I'm no expert by any means.

=====================================

</details>

## An example of how this manifests on different charts

**tl;dr**: I take a few images from [my own site](https://data-visualisation.pages.dev/HistogramMaker) to show how this looks, including the line graph which I've not actually got live yet in my work.

I've linked directly to the histogram page because it's probably the best one for demonstrating the ideas and code plus it's interactive, and using "deviation" will show the alpha level opacity which `hsla` provides.

<details>
  <summary>HSL, how is works, etc. - large section with images</summary>
  
=====================================

### Line graph

Seen as that's the subject, that's what I'll show first

<img width="1257" alt="Screenshot 2024-08-25 at 22 41 57" src="https://github.com/user-attachments/assets/b096ffa7-43ed-469e-8d85-4bb05801282a">

### Histogram

Though I'm also going to show this - IMO, it works very well with Histograms (therefore also Marimekko, Stacked ...

#### The first example doesn't need to toggle, spreading the colour works: 360 / 3 = quiet enough difference.
<img width="1308" alt="Screenshot 2024-08-25 at 22 50 25" src="https://github.com/user-attachments/assets/b443bfdb-5100-4e53-9936-f017d469941f">

#### This set of two show with a force toggle and without - the code section will cover the toggle

<img width="1133" alt="Screenshot 2024-08-25 at 22 44 49" src="https://github.com/user-attachments/assets/3ab63a42-7d2c-4d35-9f5b-b7466a179e7f">
<img width="1139" alt="Screenshot 2024-08-25 at 22 44 37" src="https://github.com/user-attachments/assets/3b26a125-5508-49ae-be02-89499a98f825">

=====================================

</details>

## Code

A couple of [functions from my work](https://github.com/toni-sharpe/data-visualisation/blob/main/gui/src/util/UtilHue/UtilHue.js) are defined, cleaned of some cruft and annotated. I use this in quite a few places.

<details><summary>Code</summary>

```
function calcContrastToggle({ i, total, useHueContrastToggle }) {
  // small sets don't need a toggle, see the first Histogram example
  // there are even cases where applying the toggle is detrimental 
  if (total <= CONTRAST_TOGGLE_MINIMUM) {
    return 0
  }
  
  // possibly a five minute job
  return (i % 2) * 180
}
```

```
export function calcAccessibleHue() {
  return function({ i, total, aLevel = 1 }) {
    // i just moves us round the wheel with the total defining how far on each step
    const hueCalced = calcHue({ i, total })

    // uses the above function
    const contrastToggleCalced = calcContrastToggle({ i, total, useHueContrastToggle })
    
    // hue + toggle gives us the highest contrast hue
    const hue = hueCalced + contrastToggleCalced
    
    // this line just ensures that we stay inside a circle 0 .. 360
    const hueCircled = hue > 360 ? hue % 360 : hue

    // and format, in my code the style tag is used
    return `hsla(${hueCircled} 80% 50% / ${aLevel})`
  }
}
```

</details> 

What's notable about the code, apart from how simple it is, is that the colour choice is basic arithmetic and very clear to understand for the reader. This relates back to the earlier section about manipulation and being more human friendly than HEX (or RGB for that matter). It matches how artists and designers think so the developer is using a language that's understood by a broader set of colleagues who don't need to be technical.

Then if you're confronted by more challenging colour logic, like what might lie under support of the various colour interpretations people have, my hypothesis would be that the task would be easier using this tool than direct HEX values.

**Final note**: there are numerous accessibility problems with "just colour" solutions, but that's a separate topic.

**Comment by CGiattino** [comment]
I'd originally sent [this message on Slack](https://owid.slack.com/archives/C04VBQ00HQR/p1726227391796539?thread_ts=1724419648.303409&cid=C04VBQ00HQR) on 13 Sep. Just copying here so it's kept together with this issue.

I think there are two related things here:
1. does the grapher select colors the way we want it to, and do this consistently?
2. is our color palette as distinct as possible? particularly in the sequence of colors that grapher chooses

I wanted to add that I experienced a similar thing today, with grapher and similar (to my eye at least!) line colors being selected.

I’m curious if others see these colors as more distinct than I do? this could all just be me!

- I went to post Este’s DI on Chile solar on social media, but I noticed the line colors looked fairly similar to my eye — I see 4 broadly reddish ones (Chile, Spain, China, US) out of 6 lines

![solar-energy_di-image (1)](https://github.com/user-attachments/assets/f62e66af-ca4b-46e3-bbf1-af0429f11382)

- so I went to [the grapher chart](https://ourworldindata.org/grapher/solar-share-energy?tab=chart&time=2000..latest) to see if I could make those a bit more distinct
- I deselected the default countries and added in the ones from the DI, and it’s perhaps a bit better, but I still got what seem to me like 4 broadly reddish colors

![solar-energy_clear-defaults-then-select (1)](https://github.com/user-attachments/assets/b71dd315-05f9-4875-87d1-d49ac2141f4e)

- I clicked refresh to see if that would improve it, as it sometimes does, and that resulted in the following. I do think the way these countries are grouped makes this seem worse than it would otherwise, but I still see 3 reddish ones and 2 greenish ones

![solar-energy_clear-select-refresh (1)](https://github.com/user-attachments/assets/3c416792-07df-4105-879f-79dc5a1e8427)

**Comment by toni-sharpe** [comment]
Thanks for sharing the Slack information, I'll make some notes on what I view shortly

**Comment by toni-sharpe** [comment]
I came back to this thinking some simulations m,ight be worth looking at. I ran them through the simulators found here at [PillStone](https://pilestone.com/). The [simulator is on this page](https://pilestone.com/pages/color-blindness-simulator-1). I ran all eight, which you'll find in the enclosed sections below. Without knowing a great deal about the medical side of this topic I took it on trust, and high google positiion, but a good case for their breakdown is that it matches StoryBook's. Parlance is copied, but the topic's got some depth. The whole green and red "classic" puts forward complexities beyond this comment.

<details><summary>Without simulation</summary>
<p>
<img width="908" alt="all-colour-options" src="https://github.com/user-attachments/assets/0c089e61-265a-4a5e-a7c1-3240f6541e44">
</p>
</details> 

<details><summary>Green: Deuter: anomaly (weak) anopia (blind)</summary>
<p>

![deutan-green](https://github.com/user-attachments/assets/26e6468e-3968-460d-8753-44f29b84fe37)

</p>
</details> 

<details><summary>Red: Protan: anomaly (weak) anopia (blind)</summary>

<p> Shows red deficiencies as simulation given the examples </p>

<p>

![protan-red](https://github.com/user-attachments/assets/933efe63-6265-457e-b960-0188dcaf171e)

</p>
</details> 

<details><summary>Blue: Tritan: anomaly (weak) anopia (blind)</summary>
<p>

![tritan-blue](https://github.com/user-attachments/assets/e4d6fa9f-da71-41e7-b568-ca886434c799)

</p>
</details> 

<details><summary>B&W & blue only</summary>
<p>

<img width="735" alt="Screenshot 2024-09-20 at 19 06 56" src="https://github.com/user-attachments/assets/fb61ed46-d89e-4e40-807a-e23c3f9e8fd9">

</p>
</details>

**Comment by lucasrodes** [comment]
To add to my previous comment, I think the current color-picking logic makes even less sense in the admin page (i.e. when we are creating a chart).

In a new chart, when choosing the entities to display, the colors picked should not be affected by whether the power user did unselections. 

My sense, however, is that the colors that the power user sees when creating a new chart vs. the colors shown when the chart is actually public don't necessarily match. That is, the colors in the latter will be based on the order of our color palette (i.e. regardless of any unselections).
